### PR TITLE
Refactors forensics to a datum

### PR DESCRIPTION
**Pull request by ghost**
there is now a forensics datum that contains like, blood, fibers, adminprints, and normal prints. everything uses it now.

No changelog because there is no visible change except in the code

Please note: /atoms still have print/fiber/blood related procs which just points to the forensics datum's procs

### CODE REVIEW

**Comment by Iamgoofball** [comment]
:+1:

Good cleanup

**Comment by MoreRobustThanYou** [comment]
@Iamgoofball ignore the massive amount of commits, this was just ripped straight from the bay forensics branch, after getting rid of the bay forensics and readding the scanner

**Comment by MoreRobustThanYou** [comment]
merge-ready now

File: code/datums/forensics/forensics.dm [hunk 1]
```diff
@@ -0,0 +1,14 @@
+//////////////////////FORENSICS DATUM
```
Comment (Cyberboss) [inline code/datums/forensics/forensics.dm]: why is it in a subfolder of datum?

File: code/datums/forensics/forensics.dm [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+	var/list/gsr
+	var/list/prints
+	var/list/hiddenprints
+	var/list/fibers
+	var/list/blood
+	var/maxSize = 5
+
+/datum/forensics/proc/enforceSize(list/templist, newthing)
+	if(templist.len >= maxSize)
+		templist.Remove(templist[1])
```
Comment (Cyberboss) [inline code/datums/forensics/forensics.dm]: -= instead of Remove

File: code/game/atoms.dm [hunk 3]
```diff
@@ -91,6 +89,9 @@
 	if(reagents)
 		qdel(reagents)
 
+	if(forensics)
+		QDEL_NULL(forensics)
```
Comment (Cyberboss) [inline code/game/atoms.dm]: just QDEL_NULL

File: code/game/atoms.dm [hunk 4]
```diff
@@ -342,30 +343,30 @@ GLOBAL_LIST_EMPTY(blood_splatter_icons)
...
-	if(!blood_DNA)	//if our list of DNA doesn't exist yet, initialise it.
-		blood_DNA = list()
-	var/old_length = blood_DNA.len
-	blood_DNA |= new_blood_dna
-	if(blood_DNA.len == old_length)
+	if(!forensics)	//if our list of DNA doesn't exist yet, initialise it.
+		forensics = new
+	var/old_length = forensics.blood.len
+	forensics.blood |= new_blood_dna
+	if(forensics.blood.len == old_length)
```
Comment (Cyberboss) [inline code/game/atoms.dm]: These three lines should be a proc on the datum

File: code/game/atoms.dm [hunk 5]
```diff
@@ -342,30 +343,30 @@ GLOBAL_LIST_EMPTY(blood_splatter_icons)
...
-	if(!blood_DNA)
-		blood_DNA = list()
-	var/old_length = blood_DNA.len
-	blood_DNA |= blood_dna
-	if(blood_DNA.len > old_length)
+	if(!forensics)
+		forensics = new
+	var/old_length = forensics.blood.len
+	forensics.blood |= blood_dna
+	if(forensics.blood.len > old_length)
```
Comment (Cyberboss) [inline code/game/atoms.dm]: These three lines should be a proc on the datum

File: code/game/atoms.dm [hunk 6]
```diff
@@ -426,8 +427,8 @@ GLOBAL_LIST_EMPTY(blood_splatter_icons)
 	return 1
 
 /atom/proc/clean_blood()
-	if(islist(blood_DNA))
-		blood_DNA = null
+	if(islist(forensics.blood))
+		forensics.blood = null
```
Comment (Cyberboss) [inline code/game/atoms.dm]: forensics proc

File: code/game/objects/effects/decals/cleanable/aliens.dm [hunk 7]
```diff
@@ -6,9 +6,14 @@
 	icon = 'icons/effects/blood.dmi'
 	icon_state = "xfloor1"
 	random_icon_states = list("xfloor1", "xfloor2", "xfloor3", "xfloor4", "xfloor5", "xfloor6", "xfloor7")
-	blood_DNA = list("UNKNOWN DNA" = "X*")
 	bloodiness = MAX_SHOE_BLOODINESS
 	blood_state = BLOOD_STATE_XENO
+	var/list/viruses = list()
+
+/obj/effect/decal/cleanable/xenoblood/Initialize(a)
+	. = ..(a)
```
Comment (Cyberboss) [inline code/game/objects/effects/decals/cleanable/aliens.dm]: `a` should be `mapload` or not exist at all

File: code/game/objects/effects/decals/cleanable/aliens.dm [hunk 8]
```diff
@@ -6,9 +6,14 @@
 	icon = 'icons/effects/blood.dmi'
 	icon_state = "xfloor1"
 	random_icon_states = list("xfloor1", "xfloor2", "xfloor3", "xfloor4", "xfloor5", "xfloor6", "xfloor7")
-	blood_DNA = list("UNKNOWN DNA" = "X*")
 	bloodiness = MAX_SHOE_BLOODINESS
 	blood_state = BLOOD_STATE_XENO
+	var/list/viruses = list()
```
Comment (Cyberboss) [inline code/game/objects/effects/decals/cleanable/aliens.dm]: create this list in Initialize

File: code/game/objects/effects/decals/cleanable/aliens.dm [hunk 9]
```diff
@@ -62,4 +67,7 @@
 /obj/effect/decal/cleanable/blood/xtracks
 	icon_state = "xtracks"
 	random_icon_states = null
-	blood_DNA = list("UNKNOWN DNA" = "X*")
+
+/obj/effect/decal/cleanable/blood/xtracks/Initialize(a)
+	. = ..(a)
```
Comment (Cyberboss) [inline code/game/objects/effects/decals/cleanable/aliens.dm]: `a` should be `mapload` or not exist at all

File: code/game/objects/effects/decals/cleanable/humans.dm [hunk 10]
```diff
@@ -4,13 +4,13 @@
 	icon = 'icons/effects/blood.dmi'
 	icon_state = "floor1"
 	random_icon_states = list("floor1", "floor2", "floor3", "floor4", "floor5", "floor6", "floor7")
-	blood_DNA = list()
 	blood_state = BLOOD_STATE_HUMAN
+	var/list/viruses = list()
 	bloodiness = MAX_SHOE_BLOODINESS
```
Comment (Cyberboss) [inline code/game/objects/effects/decals/cleanable/humans.dm]: viruses should be at the bottom

File: code/game/objects/items/stacks/stack.dm [hunk 11]
```diff
@@ -304,9 +304,9 @@
 		. = ..()
 
 /obj/item/stack/proc/copy_evidences(obj/item/stack/from as obj)
-	src.blood_DNA = from.blood_DNA
-	src.fingerprints  = from.fingerprints
-	src.fingerprintshidden  = from.fingerprintshidden
+	src.forensics.blood = from.forensics.blood
+	src.forensics.prints  = from.forensics.prints
+	src.forensics.hiddenprints  = from.forensics.hiddenprints
```
Comment (Cyberboss) [inline code/game/objects/items/stacks/stack.dm]: remove the srcs

File: code/game/atoms.dm [hunk 12]
```diff
@@ -342,30 +343,30 @@ GLOBAL_LIST_EMPTY(blood_splatter_icons)
...
+	var/old_length = forensics.blood.len
+	forensics.blood |= blood_dna
+	if(forensics.blood.len > old_length)
 		return 1//some new blood DNA was added
 
 
 //to add blood from a mob onto something, and transfer their dna info
 /atom/proc/add_mob_blood(mob/living/M)
-	var/list/blood_dna = M.get_blood_dna_list()
+	var/list/blood_dna = M.forensics.blood
```
Comment (Cyberboss) [inline code/game/atoms.dm]: This entire proc should be a forensics proc

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by Cyberboss** [comment]
@MoreRobustThanYou

**Comment by MoreRobustThanYou** [comment]
I'm busy working on other things, so i'm closing this for now

**Comment by MoreRobustThanYou** [comment]
gonna start working on this again since spacepods works now

**Comment by MoreRobustThanYou** [comment]
@Cyberboss review again please?

File: code/datums/forensics/forensics.dm (hunk 1 above)
Comment (MoreRobustThanYou) [inline code/datums/forensics/forensics.dm]: organization? i dunno

Comment (Cyberboss) [inline code/datums/forensics/forensics.dm]: Move this to the top level of the datum folder

File: code/datums/forensics/forensics.dm [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+	var/list/hiddenprints
+	var/list/fibers
+	var/list/blood
+	var/maxSize = 5
+
+/datum/forensics/proc/transfer_mob_blood_dna(mob/living/L)
+	// Returns 0 if we have that blood already
+	var/new_blood_dna = blood
+	if(!new_blood_dna)
+		return 0
```
Comment (Cyberboss) [inline code/datums/forensics/forensics.dm]: FALSE

File: code/datums/forensics/forensics.dm [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+	var/old_length = blood.len
+	blood |= new_blood_dna
+	if(blood.len == old_length)
+		return 0
```
Comment (Cyberboss) [inline code/datums/forensics/forensics.dm]: FALSE

File: code/datums/forensics/forensics.dm [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+	return 1
```
Comment (Cyberboss) [inline code/datums/forensics/forensics.dm]: TRUE

File: code/datums/forensics/forensics.dm [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+/obj/proc/transfer_mob_blood_dna(mob/living/L)
+	. = forensics.transfer_mob_blood_dna(L)
+
+//to add blood dna info to the object's blood_DNA list
+/datum/forensics/proc/transfer_blood_dna(list/blood_dna)
+	var/old_length = blood.len
+	blood |= blood_dna
+	if(blood.len > old_length)
+		return 1//some new blood DNA was added
+
```
Comment (Cyberboss) [inline code/datums/forensics/forensics.dm]: TRUE and return FALSE otherwise

File: code/datums/forensics/forensics.dm [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+/obj/proc/transfer_blood_dna(list/blood_dna)
+	. = forensics.transfer_blood_dna(blood_dna)
+
+//to add blood from a mob onto something, and transfer their dna info
+/datum/forensics/proc/add_mob_blood(mob/living/M)
+	var/list/blood_dna = M.forensics.blood
+	if(!blood_dna)
+		return 0
```
Comment (Cyberboss) [inline code/datums/forensics/forensics.dm]: FALSE

File: code/datums/forensics/forensics.dm [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+	return add_blood(blood_dna)
+
+/obj/proc/add_mob_blood(mob/living/M)
+	. = forensics.add_mob_blood(M)
+
+
+/datum/forensics/proc/add_blood(list/blood_dna)
+	return 0
```
Comment (Cyberboss) [inline code/datums/forensics/forensics.dm]: Why is this empty?

File: code/datums/forensics/forensics.dm [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+/obj/proc/add_blood(list/blood_dna)
+	. = forensics.transfer_blood_dna(blood_dna)
+
+//to add blood onto something, with blood dna info to include.
+
+
+/obj/item/add_blood(list/blood_dna)
+	var/blood_count = !forensics.blood ? 0 : forensics.blood.len
+	if(!..())
+		return 0
```
Comment (Cyberboss) [inline code/datums/forensics/forensics.dm]: FALSE

File: code/datums/forensics/forensics.dm [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+	if(!blood_count)//apply the blood-splatter overlay if it isn't already in there
+		add_blood_overlay()
+	return 1 //we applied blood to the item
```
Comment (Cyberboss) [inline code/datums/forensics/forensics.dm]: TRUE

File: code/datums/forensics/forensics.dm [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+
+/datum/forensics/proc/clean_blood()
+	if(islist(blood))
+		blood = null
+		return 1
+
```
Comment (Cyberboss) [inline code/datums/forensics/forensics.dm]: TRUE and return FALSE otherwise

File: code/game/atoms.dm [hunk 22]
```diff
@@ -263,7 +261,7 @@
 /atom/proc/examine(mob/user)
 	//This reformat names to get a/an properly working on item descriptions when they are bloody
 	var/f_name = "\a [src]."
-	if(src.blood_DNA && !istype(src, /obj/effect/decal))
+	if(forensics.blood > 0 && !istype(src, /obj/effect/decal))
```
Comment (Cyberboss) [inline code/game/atoms.dm]: .len?

File: code/game/machinery/suit_storage_unit.dm [hunk 23]
```diff
@@ -235,7 +235,7 @@
 			playsound(src, 'sound/machines/airlockclose.ogg', 25, 1)
 			for(var/obj/item/I in src) //Scorches away blood and forensic evidence, although the SSU itself is unaffected
 				I.clean_blood()
-				I.fingerprints = list()
+				I.forensics.prints = list()
```
Comment (Cyberboss) [inline code/game/machinery/suit_storage_unit.dm]: This should be a proc on the datum proc/clean_prints or something

File: code/game/atoms.dm [hunk 24]
```diff
@@ -6,12 +6,10 @@
...
 	var/list/secondary_flags
 
-	var/list/fingerprints
-	var/list/fingerprintshidden
-	var/list/blood_DNA
 	var/container_type = 0
 	var/admin_spawned = 0	//was this spawned by an admin? used for stat tracking stuff.
-	var/datum/reagents/reagents = null
+	var/datum/reagents/reagents
+	var/datum/forensics/forensics
```
Comment (Cyberboss) [inline code/game/atoms.dm]: do you ever actually initialize this somewhere?

File: code/game/gamemodes/devil/true_devil/_true_devil.dm [hunk 25]
```diff
@@ -66,7 +66,8 @@
 	//Left hand items
 	for(var/obj/item/I in held_items)
 		if(!(I.flags & ABSTRACT))
-			if(I.blood_DNA)
+
+			if(I.forensics && I.forensics.blood)
```
Comment (Cyberboss) [inline code/game/gamemodes/devil/true_devil/_true_devil.dm]: is forensics lazy or what?

File: code/modules/detectivework/detective_work.dm [hunk 26]
```diff
@@ -1,15 +1,15 @@
-//CONTAINS: Suit fibers and Detective's Scanning Computer
+//CONTAINS: Suit fibers, GSR, and Detective's Scanning Computer
```
Comment (Cyberboss) [inline code/modules/detectivework/detective_work.dm]: Undo

File: code/modules/detectivework/detective_work.dm [hunk 27]
```diff
@@ -89,30 +89,36 @@
...
+		if(!forensics.prints) //Add the list if it does not exist
+			forensics.prints = list()
 		var/full_print = md5(H.dna.uni_identity)
-		fingerprints[full_print] = full_print
+		forensics.prints[full_print] = full_print
 
 
 
 
 /atom/proc/transfer_fingerprints_to(atom/A)
```
Comment (Cyberboss) [inline code/modules/detectivework/detective_work.dm]: This should be a forensics proc

File: code/modules/detectivework/detective_work.dm [hunk 28]
```diff
@@ -89,30 +89,36 @@
...
-		A.fingerprints |= fingerprints.Copy()            //detective
-	if(fingerprintshidden)
-		A.fingerprintshidden |= fingerprintshidden.Copy()    //admin
+	if(forensics && forensics.prints)
+		A.forensics.prints |= forensics.prints.Copy()            //detective
+	if(forensics.hiddenprints)
+		A.forensics.hiddenprints |= forensics.hiddenprints.Copy()    //admin
 	A.fingerprintslast = fingerprintslast
+
+/atom/proc/add_gsr(mob/living/carbon/human/M, gsrtype)
```
Comment (Cyberboss) [inline code/modules/detectivework/detective_work.dm]: I thought we weren't doing this?

File: code/modules/detectivework/scanner.dm [hunk 29]
```diff
@@ -76,11 +76,13 @@
 
 		// Start gathering
 
-		if(A.blood_DNA && A.blood_DNA.len)
-			blood = A.blood_DNA.Copy()
 
-		if(A.suit_fibers && A.suit_fibers.len)
-			fibers = A.suit_fibers.Copy()
+
+		if(A.forensics && LAZYLEN(A.forensics.blood))
```
Comment (Cyberboss) [inline code/modules/detectivework/scanner.dm]: lazy or not?

File: code/modules/detectivework/scanner.dm [hunk 30]
```diff
@@ -76,11 +76,13 @@
...
-		if(A.blood_DNA && A.blood_DNA.len)
-			blood = A.blood_DNA.Copy()
 
-		if(A.suit_fibers && A.suit_fibers.len)
-			fibers = A.suit_fibers.Copy()
+
+		if(A.forensics && LAZYLEN(A.forensics.blood))
+			blood = A.forensics.blood.Copy()
+
+		if(A.forensics && LAZYLEN(A.forensics.fibers))
```
Comment (Cyberboss) [inline code/modules/detectivework/scanner.dm]: lazy or not?

File: code/modules/detectivework/scanner.dm [hunk 31]
```diff
@@ -90,8 +92,8 @@
 
 		else if(!ismob(A))
 
-			if(A.fingerprints && A.fingerprints.len)
-				fingerprints = A.fingerprints.Copy()
+			if(A.forensics && LAZYLEN(A.forensics.prints))
```
Comment (Cyberboss) [inline code/modules/detectivework/scanner.dm]: lazy or not?

File: code/modules/detectivework/scanner.dm [hunk 32]
```diff
@@ -113,31 +115,31 @@
 		add_log("<B>[worldtime2text()][get_timestamp()] - [target_name]</B>", 0)
 
 		// Fingerprints
-		if(fingerprints && fingerprints.len)
+		if(forensics && LAZYLEN(forensics.prints))
```
Comment (Cyberboss) [inline code/modules/detectivework/scanner.dm]: lazy or not?

File: code/modules/detectivework/scanner.dm [hunk 33]
```diff
@@ -113,31 +115,31 @@
...
 			sleep(30)
 			add_log("<span class='info'><B>Prints:</B></span>")
 			for(var/finger in fingerprints)
 				add_log("[finger]")
 			found_something = 1
 
 		// Blood
-		if (blood && blood.len)
+		if (forensics && LAZYLEN(forensics.blood))
```
Comment (Cyberboss) [inline code/modules/detectivework/scanner.dm]: lazy or not?

File: code/modules/detectivework/scanner.dm [hunk 34]
```diff
@@ -113,31 +115,31 @@
...
 			sleep(30)
 			add_log("<span class='info'><B>Blood:</B></span>")
 			found_something = 1
 			for(var/B in blood)
 				add_log("Type: <font color='red'>[blood[B]]</font> DNA: <font color='red'>[B]</font>")
 
 		//Fibers
-		if(fibers && fibers.len)
+		if(forensics && LAZYLEN(forensics.fibers))
```
Comment (Cyberboss) [inline code/modules/detectivework/scanner.dm]: lazy or not?

File: code/modules/mob/living/carbon/examine.dm [hunk 35]
```diff
@@ -19,7 +19,7 @@
 
 	for(var/obj/item/I in held_items)
 		if(!(I.flags & ABSTRACT))
-			if(I.blood_DNA)
+			if(I.forensics && I.forensics.blood)
```
Comment (Cyberboss) [inline code/modules/mob/living/carbon/examine.dm]: More of the same for the rest of the review...

**Review by Cyberboss — DISMISSED** [review]
You're transforming this into a component before it's merged, but let's fix the shitcode first

File: code/modules/detectivework/detective_work.dm (hunk 28 above)
Comment (MoreRobustThanYou) [inline code/modules/detectivework/detective_work.dm]: Oh I forgot

File: code/game/atoms.dm (hunk 24 above)
Comment (MoreRobustThanYou) [inline code/game/atoms.dm]: CheckParts?

File: code/modules/detectivework/detective_work.dm (hunk 28 above)
Comment (MoreRobustThanYou) [inline code/modules/detectivework/detective_work.dm]: Ok, fixed.

File: code/modules/detectivework/detective_work.dm (hunk 27 above)
Comment (MoreRobustThanYou) [inline code/modules/detectivework/detective_work.dm]: Done!

File: code/game/machinery/suit_storage_unit.dm (hunk 23 above)
Comment (MoreRobustThanYou) [inline code/game/machinery/suit_storage_unit.dm]: Done!

**Comment by MoreRobustThanYou** [comment]
@Cyberboss should be ok now

File: code/game/machinery/suit_storage_unit.dm (hunk 23 above)
Comment (Cyberboss) [inline code/game/machinery/suit_storage_unit.dm]: it's still here...

File: code/game/atoms.dm (hunk 24 above)
Comment (Cyberboss) [inline code/game/atoms.dm]: CheckParts doesnt exist on all atoms

Comment (MoreRobustThanYou) [inline code/game/atoms.dm]: yes but it exists in /atom itself.

File: code/game/machinery/suit_storage_unit.dm (hunk 23 above)
Comment (MoreRobustThanYou) [inline code/game/machinery/suit_storage_unit.dm]: added clean_prints

File: code/game/atoms.dm (hunk 24 above)
Comment (MoreRobustThanYou) [inline code/game/atoms.dm]: Ok, it should now initalize in, well initalize

**Comment by MoreRobustThanYou** [comment]
@Cyberboss this ready now?

**Comment by MoreRobustThanYou** [comment]
WTF DID GIT JUST AUTO SOLVE A CONFLICT

**Comment by MoreRobustThanYou** [comment]
@Cyberboss re-review?

File: code/game/atoms.dm [hunk 36]
```diff
@@ -161,6 +162,9 @@
 				reagents = new()
 			reagents.reagent_list.Add(A)
 			reagents.conditional_update()
+		if(istype(A, /datum/forensics))
+			if(!forensics)
+				forensics = new
```
Comment (kevinz000) [inline code/game/atoms.dm]: isn't this already done in initialize

Comment (MoreRobustThanYou) [inline code/game/atoms.dm]: yes, but i think CheckParts is to ensure all the vars exist and are correct

**Comment by MoreRobustThanYou** [comment]
Uhhh, travis, you drunk?

**Comment by XDTM** [comment]
I haven't looked indepth, but could viruses potentially piggyback on this system, so they infect people in contact with blood?

File: code/modules/mob/living/simple_animal/friendly/drone/_drone.dm [hunk 37]
```diff
@@ -176,21 +176,21 @@
...
-			if(I.blood_DNA)
+			if(LAZYLEN(I.forensics.blood))
 				msg += "<span class='warning'>It has [icon2html(I, user)] [I.gender==PLURAL?"some":"a"] blood-stained [I.name] in its [get_held_index_name(get_held_index_of_item(I))]!</span>\n"
 			else
 				msg += "It has [icon2html(I, user)] \a [I] in its [get_held_index_name(get_held_index_of_item(I))].\n"
 
 	//Internal storage
-	if(internal_storage && !(internal_storage.flags_1&ABSTRACT_1))
-		if(internal_storage.blood_DNA)
+  if(internal_storage && !(internal_storage.flags_1&ABSTRACT_1))
```
Comment (optimumtact) [inline code/modules/mob/living/simple_animal/friendly/drone/_drone.dm]: code/modules/mob/living/simple_animal/friendly/drone/_drone.dm:185:warning: if: if statement has no effect

I'm not sure exactly why it's warning on this line tbh, something about the flags is making it confused, it's possible it thinks you're trying to do (internal_storage && !internal_storage)

try
```
var/flagcheck = internal_storage.flags_1&ABSTRACT_1
if(internal_storage && flagcheck)
```

File: code/modules/mob/living/simple_animal/friendly/drone/_drone.dm [hunk 38]
```diff
@@ -176,21 +176,22 @@
...
+			if(LAZYLEN(I.forensics.blood))
 				msg += "<span class='warning'>It has [icon2html(I, user)] [I.gender==PLURAL?"some":"a"] blood-stained [I.name] in its [get_held_index_name(get_held_index_of_item(I))]!</span>\n"
 			else
 				msg += "It has [icon2html(I, user)] \a [I] in its [get_held_index_name(get_held_index_of_item(I))].\n"
 
 	//Internal storage
-	if(internal_storage && !(internal_storage.flags_1&ABSTRACT_1))
-		if(internal_storage.blood_DNA)
+	var/flagcheck = internal_storage.flags_1&ABSTRACT_1
+	if(internal_storage && flagcheck)
```
Comment (optimumtact) [inline code/modules/mob/living/simple_animal/friendly/drone/_drone.dm]: this should be !flagcheck, otherwise you've inverted the previous setting

**Comment by MoreRobustThanYou** [comment]
why is it feel like this is just gonna sit here and nobody will touch it

**Comment by Cyberboss** [comment]
r.e. the last time I tried to refactor forensics

**Comment by MoreRobustThanYou** [comment]
nice

**Comment by optimumtact** [comment]
I'm going to print out all the bugs this causes and throw them at your head  in a stack

**Comment by MoreRobustThanYou** [comment]
I love you too

**Comment by Cyberboss** [comment]
You've doomed us all

**Comment by ShizCalev** [comment]
holy fuck runtimes out the ass

**Pull request merged**